src/borg/testsuite/compress.py

```diff
@@ -221,51 +178,40 @@ def test_obfuscate_meta():
     assert 0 < psize < 100
-    assert csize - psize >= 0  # there is a obfuscation trailer
+    assert csize - psize >= 0  # there is an obfuscation trailer
     trailer = compressed[psize:]
     assert not trailer or set(trailer) == {0}  # trailer is all-zero-bytes
 
+@pytest.mark.parametrize(
+    "c_type, c_name, c_levels",
+    [
+        (ZLIB, "zlib", ["", 0, 9, "9,invalid"]),
+        (LZMA, "lzma", ["", 0, 9, "9,invalid"]),
+        (ZSTD, "zstd", ["", 1, 22, "22,invalid"]),
+        (CNONE, "none", [""]),
+        (LZ4, 'lz4', [""]),
+        (None, None, ["invalid"]),
+    ]
+)
+def test_compression_specs(c_type, c_name, c_levels):
+    for level in c_levels:
+        # a compression level is specified
+        if isinstance(level, int):
+            cs = CompressionSpec(f"{c_name},{level}").compressor
+            assert isinstance(cs, c_type)
+            assert cs.level == level
+
+        # a compression level is not specified (default)
+        elif not level:
+            cs = CompressionSpec(c_name).compressor
+            assert isinstance(cs, c_type)
+            if c_type == ZLIB | LZMA:
+                assert cs.level == 6
+            elif c_type == ZSTD:
+                assert cs.level == 3
+
+        # a compression level is invalid
+        elif level.endswith("invalid"):
+            with pytest.raises(argparse.ArgumentTypeError):
+                CompressionSpec(f"{c_name},{level}")
 
-def test_compression_specs():
-    with pytest.raises(argparse.ArgumentTypeError):
-        CompressionSpec("")
-
-    assert isinstance(CompressionSpec("none").compressor, CNONE)
-    assert isinstance(CompressionSpec("lz4").compressor, LZ4)
-
-    zlib = CompressionSpec("zlib").compressor
-    assert isinstance(zlib, ZLIB)
-    assert zlib.level == 6
-    zlib = CompressionSpec("zlib,0").compressor
-    assert isinstance(zlib, ZLIB)
-    assert zlib.level == 0
-    zlib = CompressionSpec("zlib,9").compressor
-    assert isinstance(zlib, ZLIB)
-    assert zlib.level == 9
-    with pytest.raises(argparse.ArgumentTypeError):
-        CompressionSpec("zlib,9,invalid")
-
-    lzma = CompressionSpec("lzma").compressor
-    assert isinstance(lzma, LZMA)
-    assert lzma.level == 6
-    lzma = CompressionSpec("lzma,0").compressor
-    assert isinstance(lzma, LZMA)
-    assert lzma.level == 0
-    lzma = CompressionSpec("lzma,9").compressor
-    assert isinstance(lzma, LZMA)
-    assert lzma.level == 9
-
-    zstd = CompressionSpec("zstd").compressor
-    assert isinstance(zstd, ZSTD)
-    assert zstd.level == 3
-    zstd = CompressionSpec("zstd,1").compressor
-    assert isinstance(zstd, ZSTD)
-    assert zstd.level == 1
-    zstd = CompressionSpec("zstd,22").compressor
-    assert isinstance(zstd, ZSTD)
-    assert zstd.level == 22
-
-    with pytest.raises(argparse.ArgumentTypeError):
-        CompressionSpec("lzma,9,invalid")
-    with pytest.raises(argparse.ArgumentTypeError):
-        CompressionSpec("invalid")
```

Comment:
guess `CompressionSpec("invalid")` is missing now?

Reply:
Attempted to replicate this with the `(None, None, ["invalid"]),` parameter on line 193

Reply:
Doesn't that test with `"None,invalid"` rather than `"invalid"`.